woocommerce/google-listings-and-ads
# e2e testing: merchant getting started to Setup MC Step 1

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/google-listings-and-ads/issues/789.

This PR puts in a new foundation for e2e testing using [`@woocommerce/e2e-environment`](https://www.npmjs.com/package/@woocommerce/e2e-environment) package and adds a basic e2e test covering merchant getting started from login into WordPress to Setup MC Step 1 page.

Most of the work here was done by referring to the test implementation in WooCommerce Core and WooCommerce Admin.

#### Other technical notes:

This PR also sets the `.nvmrc` file to `lts/*`, same as [WooCommerce Admin](https://github.com/woocommerce/woocommerce-admin/blob/main/.nvmrc).

The files `globalTypes.d.ts` and `jsconfig.json` are added to enable VS Code intellisense for `puppeteer`:

![vs-code-puppeteer-intellisense](https://user-images.githubusercontent.com/417342/122965184-84c7da80-d3ba-11eb-8119-7f1a2457a0f4.png)

`@woocommerce/api` package is added even though we are not using it, because it is a peerDependency of `@woocommerce/e2e-utils` package.

ESLint now works in the `tests` folder. This is done by removing `tests` in `.eslintignore` file.

`wp-env` related things are removed from the code base.

I tried to setup e2e to run [`@woocommerce/e2e-core-tests`](https://www.npmjs.com/package/@woocommerce/e2e-core-tests), specifically the `runSetupOnboardingTests` tests. However, some tests failed, so I'm not including them.

### Screenshots:

Demo video with my voice, showing both the headless testing mode and the e2e-dev mode: (https://d.pr/v/hGmMIQ)

https://user-images.githubusercontent.com/417342/122814970-7967a700-d307-11eb-9dac-403e3e9c2dac.mov

### Detailed test instructions:

Pre-condition: make sure you have docker desktop installed and running in your machine.

1. Checkout this branch. 
2. Make sure you are running LTS version of node.
3. Make sure you've cleaned your `/vendor` from Guzzle problems `rm -rf vendor && composer install`
3. Run `npm install` in your console.
4. Run `npm run docker:up`. This will auto download images and run e2e site in docker container.
5. Run `npm run test:e2e` to run the tests in headless mode. The test should pass.
6. Run `npm run test:e2e-dev` to run the tests in dev mode. This will open up chromium browser and test will run automatically in the browser. The test should pass.
7. Optionally, if you want to clean up / throwaway the docker container, run `npm run docker:down`. This will remove the containers and image.

Test for VS Code developer experience:

1. Open VS Code. 
2. Navigate to `get-started.test.js` file.
3. Use your mouse to hover over variables like `page` and methods like `hover` and `click`. There should be a popup that tells you what it does.
4. On an empty line, type `await page.`, there should be a list of method suggestions showing up.

### Changelog entry

Adds e2e testing covering merchant getting started to Setup MC page 1.
